Activities/ManagePersonsPermissionCheckerTask.cs

```diff
@@ -0,0 +1,74 @@
+using Microsoft.Extensions.Localization;
+using OrchardCore.Security.Services;
+using OrchardCore.Users.Models;
+using OrchardCore.Users.Services;
+using OrchardCore.Workflows.Abstractions.Models;
+using OrchardCore.Workflows.Activities;
+using OrchardCore.Workflows.Models;
+using OrchardCore.Workflows.Services;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Lombiq.TrainingDemo.Activities;
+
+// A simple workflow activity task that accepts an username as a TextField input and checks whether the user has
+// ManagePersons Permission or not.
+public class ManagePersonsPermissionCheckerTask : TaskActivity
+{
+    private readonly IStringLocalizer S;
+
+    private readonly IWorkflowExpressionEvaluator _expressionEvaluator;
+
+    private readonly IUserService _userService;
+
+    private readonly IRoleService _roleService;
+
+
+    public ManagePersonsPermissionCheckerTask(
+        IUserService userService,
+        IWorkflowExpressionEvaluator expressionEvaluator,
+        IRoleService roleService,
+        IStringLocalizer<ManagePersonsPermissionCheckerTask> localizer)
+    {
+        _expressionEvaluator = expressionEvaluator;
+        S = localizer;
+        _userService = userService;
+        _roleService = roleService;
+    }
+
+    public override string Name => nameof(ManagePersonsPermissionCheckerTask);
+
+    public override LocalizedString DisplayText => S["Manage Persons Permission Checker Task"];
+
+    public override LocalizedString Category => S["User"];
+    public WorkflowExpression<string> UserName
+    {
+        get => GetProperty(() => new WorkflowExpression<string>());
+        set => SetProperty(value);
+    }
+
+    public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContext workflowContext, ActivityContext activityContext) =>
+        Outcomes(S["HasPermission"], S["NoPermission"]);
+
+    public override async Task<ActivityExecutionResult> ExecuteAsync(WorkflowExecutionContext workflowContext, ActivityContext activityContext)
+    {
+        var userName = await _expressionEvaluator.EvaluateAsync(UserName, workflowContext, null);
+        User user = (User)await _userService.GetUserAsync(userName);
+
+        if (user != null)
+        {
+            var permissions = user.RoleNames
+                .Select(async roleName => await _roleService.GetRoleClaimsAsync(roleName))
+                .SelectMany(claim => claim.Result)
+                .Where(claim => "Permission".Equals(claim.Type))
+                .Select(claim => claim.Value)
+                .Distinct()
+                .ToList();
```

Comment:
Instead of checking for the permission in a role, you can use `_authorizationService.AuthorizeAsync(user, permission);` where you use `IAuthorizationService` and the `Permission` object from `PersonPermissions`.

Reply:
Please correct me if I am wrong, but  _`authorizationService.AuthorizeAsync(user, permission); ` expects the user as a ClaimsPrincipal type. The issue is that I am only able to get the currently authenticated user as a ClaimPrincipal, but the activity receives a username as input and therefore I have to find the user by its name and check its permissions. As I see the Authorization Service does not allow this. What do you think?

Reply:
Indeed, but you can create a `ClaimsPrincipal` with `IUserService.CreatePrincipalAsync()`.

Reply:
Good point, thank you!